README.markdown
# Ogone PHP SDK #

This library allows you to easily implement an [Ogone](http://ogone.com) integration into your project.
It provides the necessary components to complete a correct payment flow with the [Ogone](http://ogone.com) platform.

Requirements:

- PHP 5.3
- network connection between your webserver and the Ogone platform


## Overview ##

The library complies to the [PSR-0 standard](http://groups.google.com/group/php-standards/web/psr-0-final-proposal),
so it can be autoloaded using PSR-0 classloaders like the one in Symfony2. See autoload.php for an example.

- Create an EcommercePaymentRequest or CreateAliasRequest, containing all the info needed by Ogone.
- Generate  a form
- Submit it to Ogone (client side)
- Receive a PaymentResponse back from Ogone (as a HTTP Request)

Both EcommercePaymentRequest, CreateAliasRequest and PaymentResponse are authenticated by comparing the SHA sign, which is a hash of the parameters and a secret passphrase. You can create the hash using a ShaComposer.

The library also allows:
- Fetching order information via Ogone API using DirectLinkQueryRequest
- Executing maintenance request via Ogone API using DirectLinkMaintenanceRequest

# SHA Composers #

Ogone provides 2 methods to generate a SHA sign:

- "Main parameters only"

  ![Main parameters only](http://github.com/marlon-be/marlon-ogone/raw/master/documentation/images/ogone_security_legacy.png)

  Implementation using this library is trivial:

```php
  <?php
	use Ogone\ShaComposer\LegacyShaComposer;
	$shaComposer = new LegacyShaComposer($passphrase);
```

- "Each parameter followed by the passphrase"

  ![Each parameter followed by the passphrase](http://github.com/marlon-be/marlon-ogone/raw/master/documentation/images/ogone_security_allparameters_sha1_utf8.png)

  Implementation using this library is trivial:

```php
  	<?php
	use Ogone\ShaComposer\AllParametersShaComposer;
	$shaComposer = new AllParametersShaComposer($passphrase);
```

This library currently supports both the legacy method "Main parameters only" and the new method "Each parameter followed by the passphrase". Either can be used with SHA-1 (default), SHA-256 or SHA-512 encryption.

# EcommercePaymentRequest and FormGenerator #

```php
	<?php
	use Ogone\Passphrase;
	use Ogone\Ecommerce\EcommercePaymentRequest;
	use Ogone\ShaComposer\AllParametersShaComposer;
	use Ogone\FormGenerator\SimpleFormGenerator;

	$passphrase = new Passphrase('my-sha-in-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new ShaInParameterFilter); //optional

	$ecommercePaymentRequest = new EcommercePaymentRequest($shaComposer);

	// Optionally set Ogone uri, defaults to TEST account
	//$ecommercePaymentRequest->setOgoneUri(EcommercePaymentRequest::PRODUCTION);

	// Set various params:
	$ecommercePaymentRequest->setOrderid('123456');
	$ecommercePaymentRequest->setAmount(150); // in cents
	$ecommercePaymentRequest->setCurrency('EUR');
	// ...

	$ecommercePaymentRequest->validate();

	$formGenerator = new SimpleFormGenerator;
	$html = $formGenerator->render($ecommercePaymentRequest);
	// Or use your own generator. Or pass $ecommercePaymentRequest to a view
```

# CreateAliasRequest #

```php
	<?php

	use Ogone\Passphrase;
	use Ogone\DirectLink\CreateAliasRequest;
	use Ogone\ParameterFilter\AliasShaInParameterFilter;
    use Ogone\ShaComposer\AllParametersShaComposer;
	use Ogone\DirectLink\Alias;

	$passphrase = new Passphrase('my-sha-in-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new AliasShaInParameterFilter);

	$createAliasRequest = new CreateAliasRequest($shaComposer);

	// Optionally set Ogone uri, defaults to TEST account
	// $createAliasRequest->setOgoneUri(CreateAliasRequest::PRODUCTION);

	// set required params
	$createAliasRequest->setPspid('123456');
	$createAliasRequest->setAccepturl('http://example.com/accept');
	$createAliasRequest->setExceptionurl('http://example.com/exception');

	// set optional alias, if empty, Ogone creates one
	$alias = new Alias('customer_123');
	$createAliasRequest->setAlias($alias);

	$createAliasRequest->validate();

	// Now pass $createAliasRequest to a view to build a custom form, you have access to
	// $createAliasRequest->getOgoneUri(), $createAliasRequest->getParameters() and $createAliasRequest->getShaSign()
	// Be sure to add the required fields CN (Card holder's name), CARDNO (Card/account number), ED (Expiry date (MMYY)), CVC (Card Verification Code)
	// and the SHASIGN
```

# DirectLinkPaymentRequest #

```php
	<?php

	use Ogone\DirectLink\DirectLinkPaymentRequest;
	use Ogone\Passphrase;
	use Ogone\ShaComposer\AllParametersShaComposer;
	use Ogone\DirectLink\Alias;

	$passphrase = new Passphrase('my-sha-in-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new ShaInParameterFilter); //optional

	$directLinkRequest = new DirectLinkPaymentRequest($shaComposer);
	$directLinkRequest->setOrderid('order_1234');

	$alias = new Alias('customer_123');
	$directLinkRequest->setAlias($alias);
	$directLinkRequest->setPspid('123456');
	$directLinkRequest->setUserId('ogone-api-user');
	$directLinkRequest->setPassword('ogone-api-password');
	$directLinkRequest->setAmount(100);
	$directLinkRequest->setCurrency('EUR');
	$directLinkRequest->validate();

	// now create a url to be posted to Ogone
	// you have access to $directLinkRequest->toArray(), $directLinkRequest->getOgoneUri() and directLinkRequest->getShaSign()
```

# DirectLinkQueryRequest #

```php
	<?php

	use Ogone\DirectLink\DirectLinkQueryRequest;
	use Ogone\Passphrase;
	use Ogone\ShaComposer\AllParametersShaComposer;
	use Ogone\DirectLink\Alias;

	$passphrase = new Passphrase('my-sha-in-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new ShaInParameterFilter); //optional

	$directLinkRequest = new DirectLinkQueryRequest($shaComposer);
	$directLinkRequest->setPspid('123456');
	$directLinkRequest->setUserId('ogone-api-user');
	$directLinkRequest->setPassword('ogone-api-password');
	$directLinkRequest->setPayId('order_1234');
	$directLinkRequest->validate();

	// now create a url to be posted to Ogone
	// you have access to $directLinkRequest->toArray(), $directLinkRequest->getOgoneUri() and directLinkRequest->getShaSign()
```

# DirectLinkMaintenanceRequest #

```php
	<?php

	use Ogone\DirectLink\DirectLinkMaintenanceRequest;
	use Ogone\DirectLink\MaintenanceOperation;
	use Ogone\Passphrase;
	use Ogone\ShaComposer\AllParametersShaComposer;
	use Ogone\DirectLink\Alias;

	$passphrase = new Passphrase('my-sha-in-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new ShaInParameterFilter); //optional

	$directLinkRequest = new DirectLinkMaintenanceRequest($shaComposer);
	$directLinkRequest->setPspid('123456');
	$directLinkRequest->setUserId('ogone-api-user');
	$directLinkRequest->setPassword('ogone-api-password');
	$directLinkRequest->setPayId('order_1234');
	$directLinkRequest->setOperation(new MaintenanceOperation(MaintenanceOperation::OPERATION_AUTHORISATION_RENEW));

	$directLinkRequest->validate();

	// now create a url to be posted to Ogone
	// you have access to $directLinkRequest->toArray(), $directLinkRequest->getOgoneUri() and directLinkRequest->getShaSign()
```

# EcommercePaymentResponse #

```php
  	<?php

	use Ogone\Ecommerce\EcommercePaymentResponse;
	use Ogone\ShaComposer\AllParametersShaComposer;

	// ...

	$ecommercePaymentResponse = new EcommercePaymentResponse($_REQUEST);

	$passphrase = new Passphrase('my-sha-out-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new ShaOutParameterFilter); //optional

	if($ecommercePaymentResponse->isValid($shaComposer) && $ecommercePaymentResponse->isSuccessful()) {
		// handle payment confirmation
	}
	else {
		// perform logic when the validation fails
	}
```

# CreateAliasResponse #

```php
  	<?php

	use Ogone\DirectLink\CreateAliasResponse;
	use Ogone\ShaComposer\AllParametersShaComposer;

	// ...

	$createAliasResponse = new CreateAliasResponse($_REQUEST);

	$passphrase = new Passphrase('my-sha-out-passphrase-defined-in-ogone-interface');
	$shaComposer = new AllParametersShaComposer($passphrase);
	$shaComposer->addParameterFilter(new ShaOutParameterFilter); //optional

	if($createAliasResponse->isValid($shaComposer) && $createAliasResponse->isSuccessful()) {
		// Alias creation is succesful, get the Alias object
		$alias = $createAliasResponse->getAlias();
	}
	else {
		// validation failed, retry?
	}
```

# DirectLinkPaymentResponse #

As the DirectLink payment gets an instant feedback from the server (and no async response) we don't use the SHA validation.

```php
	<?php

	use Ogone\DirectLink\DirectLinkPaymentResponse;

	$directLinkResponse = new DirectLinkPaymentResponse('ogone-direct-link-result-as-xml');

	if($directLinkResponse->isSuccessful()) {
    	// handle payment confirmation
	} else {
    	// perform logic when the validation fails
	}
```



# Parameter filters #
ParameterFilters are used to filter the provided parameters.
Both ShaIn- and ShaOutParameterFilters are provided and are based on the parameter lists defined in the Ogone documentation.
Parameter filtering is optional, but we recommend using them to enforce expected parameters.


CHANGELOG.md
# Changelog
## [3.0.1] - 2021-03-01
### Added
- Add `Device` parameter for Bancontact

### Changed
- Update `owneraddress` length (35)
- Change size validation of `ownertown` field (40)

## [3.0.0] - 2020-11-03
- Use mb_strlen for length checking

## [2.0.2] - 2020-10-26
### Changed
- Remove length limitions of address fields
- Update OwnerAddress field limitions
- Remove Ogone Uris validation

## [2.0.1] - 2020-09-22
### Fixed
- Issue `setOperation`: operation can be null


lib/Ogone/Validator.php
<?php

namespace Ogone;

use InvalidArgumentException;

class Validator
{
    /**
     * Types
     */
    const TYPE_ALPHANUMERIC = 'AN';
    const TYPE_NUMERIC = 'N';
    const TYPE_EMAIL = 'EMAIL';

    /**
     * Allowed Languages
     * @var array
     */
    protected $allowedLanguages = [
        'en_US',
        'cs_CZ',
        'de_DE',
        'dk_DK',
        'el_GR',
        'es_ES',
        'fr_FR',
        'it_IT',
        'ja_JP',
        'nl_BE',
        'nl_NL',
        'no_NO',
        'pl_PL',
        'pt_PT',
        'ru_RU',
        'se_SE',
        'sk_SK',
        'tr_TR',
    ];

    /**
     * Allowed Currencies.
     *
     * @var array
     */
    protected $allowedCurrencies = [
        'AED',
        'ANG',
        'ARS',
        'AUD',
        'AWG',
        'BGN',
        'BRL',
        'BYR',
        'CAD',
        'CHF',
        'CNY',
        'CZK',
        'DKK',
        'EEK',
        'EGP',
        'EUR',
        'GBP',
        'GEL',
        'HKD',
        'HRK',
        'HUF',
        'ILS',
        'ISK',
        'JPY',
        'KRW',
        'LTL',
        'LVL',
        'MAD',
        'MXN',
        'MYR',
        'NOK',
        'NZD',
        'PLN',
        'RON',
        'RUB',
        'SEK',
        'SGD',
        'SKK',
        'THB',
        'TRY',
        'UAH',
        'USD',
        'XAF',
        'XOF',
        'XPF',
        'ZAR'
    ];

    /**
     * Fields
     * @var array
     */
    protected static $fields = [
        // 2.1 Standard Ingenico ePayments fields
        'pspid' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 30,
        ],
        'orderid' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 40,
        ],
        'amount' => [
            'format' => self::TYPE_NUMERIC,
            // Any size
        ],
        'currency' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 3,
        ],
        'language' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 5,
        ],
        'operation' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 3,
            // @todo null was added as exception for Bancontact, Aurore
            'values' => ['RES', 'SAL', 'REN', 'DEL', 'DES', 'SAS', 'RFD', 'RFS', null]
        ],

        // 2.2 Invoicing and delivery data
        // https://epayments-support.ingenico.com/en/integration/all-sales-channels/integrate-with-directlink-server-to-server/guide
        // https://payment-services.ingenico.com/int/en/ogone/support/guides/integration%20guides/klarna
        'owneraddress' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 35,
            // @todo (Klarna accepts max. 35)
        ],
        'ownerzip' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'email' => [
            'format' => self::TYPE_EMAIL,
            'size' => 50,
        ],
        'ownertown' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 40,
        ],
        'ownertelno' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 20,
            // @todo (AfterPay accepts max. 10)
        ],
        'cuid' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 50,
        ],
        'ecom_consumer_gender' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 1,
            'vales' => ['m', 'f', 'M', 'F']
        ],
        'ecom_billto_postal_city' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 25,
        ],
        'ecom_billto_postal_countrycode' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 2,
        ],
        'ecom_billto_postal_county' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 25,
        ],
        'ecom_billto_postal_name_first' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_billto_postal_name_last' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_billto_postal_postalcode' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'ecom_billto_postal_street_line1' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_billto_postal_street_line2' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_billto_postal_street_number' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'ecom_shipto_company' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 50,
        ],
        'ecom_shipto_dob' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
            //@todo (format dd/MM/yyyy)
        ],
        'ecom_shipto_online_email' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 50,
        ],
        'ecom_shipto_postal_city' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 40,
            // @todo Afterpay: 25
        ],
        'ecom_shipto_postal_countrycode' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 2,
        ],
        'ecom_shipto_county' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 25,
        ],
        'ecom_shipto_postal_name_first' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_shipto_postal_name_last' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_shipto_postal_name_prefix' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'ecom_shipto_postal_postalcode' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'ecom_shipto_postal_state' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 2,
        ],
        'ecom_shipto_postal_street_line1' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_shipto_postal_street_line2' => [
            'format' => self::TYPE_ALPHANUMERIC,
            //'size' => 35,
        ],
        'ecom_shipto_postal_street_number' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'ecom_shipto_telecom_fax_number' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 20,
        ],
        'ecom_shipto_telecom_phone_number' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 20,
        ],
        'ownercty' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 2,
        ],
        'ordershipmeth' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 25,
        ],
        'ordershipcost' => [
            'format' => self::TYPE_NUMERIC,
        ],
        'ordershiptaxcode' => [
            // @todo Numberic? Alphanumberic?
        ],

        // 2.2 Invoicing and delivery data
        // https://payment-services.ingenico.com/int/en/ogone/support/guides/integration%20guides/afterpay
        'civility' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 10,
        ],
        'datein' => [
            // @todo Format: mm/dd/yyyy hh:mm:ss
        ],
        'ordershiptax' => [
            'format' => self::TYPE_NUMERIC,
        ],
        // Extra parameters for B2B
        'ref_customerref' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 20,
        ],
        'ecom_shipto_tva' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 20,
        ],
        'costcenter' => [
            'format' => self::TYPE_ALPHANUMERIC,
            'size' => 20,
        ],

        // 2.3 Order details
        // Optional integration data: Order data ("ITEM" parameters).
        // https://payment-services.ingenico.com/int/en/ogone/support/guides/integration%20guides/additional-data/order-data
        'itemattributes*' => [
            'size' => 50,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemcategory*' => [
            'size' => 50,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemcomments*' => [
            'size' => 255,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemdesc*' => [
            'size' => 16,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemdiscount*' => [
            'size' => 10,
            'format' => self::TYPE_NUMERIC
        ],
        'itemid*' => [
            'size' => 15,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemname*' => [
            'size' => 40,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemprice*' => [
            'size' => 50,
            'format' => self::TYPE_NUMERIC
        ],
        'itemquant*' => [
            'size' => 50,
            'format' => self::TYPE_NUMERIC
        ],
        'itemquantorig*' => [
            'size' => 50,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemunitofmeasure*' => [
            'size' => 50,
            'format' => self::TYPE_ALPHANUMERIC
        ],
        'itemvat*' => [
            'size' => 50,
            'format' => self::TYPE_NUMERIC
        ],
        'itemvatcode*' => [
            'size' => 50,
            'format' => self::TYPE_NUMERIC
        ],
        'itemweight*' => [
            'size' => 10,
            'format' => self::TYPE_NUMERIC
        ],
        'taxincluded*' => [
            'size' => 1,
            //'format' => self::TYPE_NUMERIC,
            'format' => self::TYPE_ALPHANUMERIC,
            'values' => [0, 1, '']
        ],
        // @see https://epayments-support.ingenico.com/en/payment-methods/alternative-payment-methods/bancontact
        'device' => [
            'vales' => ['computer', 'tablet', 'mobile', false]
        ],
    ];



    /**
     * Field Name
     * @var string
     */
    private $fieldName;

    /**
     * Field Value
     * @var string
     */
    private $fieldValue;

    /**
     * Options of Validation
     * @var array
     */
    private $validationOptions = [];

    /**
     * Validate constructor.
     *
     * @param $key
     * @param $value
     * @param array $options
     */
    public function __construct($key, $value, array $options = [])
    {
        // Check if a string ends with a number
        // Check parameters is like ITEMID*
        $last = mb_substr($key, -1, 1, 'UTF-8');
        if (is_numeric($last) && in_array(rtrim($key, $last) . '*', array_keys(self::$fields))) {
            $key = rtrim($key, $last);
        }


        $this->fieldName = mb_strtolower($key, 'UTF-8');
        $this->fieldValue = $value;
        $this->validationOptions = $options;
    }

    /**
     * Validate
     *
     * @return bool
     * @throws InvalidArgumentException
     */
    public function validate()
    {
        if (isset(self::$fields[$this->fieldName]) && is_array(self::$fields[$this->fieldName])) {
            $rules = self::$fields[$this->fieldName];

            // Validate format
            if (isset($rules['format'])) {
                // Validate numeric
                if ($rules['format'] === self::TYPE_NUMERIC &&
                    !is_numeric($this->fieldValue)
                ) {
                    throw new InvalidArgumentException(sprintf('%s is not valid format of value %s',
                        $this->fieldName,
                        $this->fieldValue
                    ));
                }

                // Validate alphanumeric
                // @todo Verify with ctype_print($this->fieldValue)
                if ($rules['format'] === self::TYPE_ALPHANUMERIC &&
                    !is_string((string) $this->fieldValue)
                ) {
                    throw new InvalidArgumentException(sprintf('%s is not valid format of value %s',
                        $this->fieldName,
                        $this->fieldValue
                    ));
                }

                // Validate E-mail
                if ($rules['format'] === self::TYPE_EMAIL &&
                    !filter_var($this->fieldValue, FILTER_VALIDATE_EMAIL)
                ) {
                    throw new InvalidArgumentException(sprintf('%s is not valid email of value %s',
                        $this->fieldName,
                        $this->fieldValue
                    ));
                }
            }

            // Validate size
            if (isset($rules['size']) && mb_strlen($this->fieldValue, 'UTF-8') > $rules['size']) {
                throw new InvalidArgumentException(sprintf('%s is too long', $this->fieldName));
            }

            // Validate possible values
            if (isset($rules['values']) && !in_array($this->fieldValue, $rules['values'])) {
                throw new InvalidArgumentException(sprintf('%s is not valid value %s', $this->fieldName, $this->fieldValue));
            }

            // Validate Language
            if ($this->fieldName === 'language' && !in_array($this->fieldValue, $this->allowedLanguages)) {
                throw new InvalidArgumentException('Invalid language ISO code');
            }

            // Validate Currency
            if ($this->fieldName === 'currency' && !in_array($this->fieldValue, $this->allowedCurrencies)) {
                throw new InvalidArgumentException('Unknown currency');
            }
        }

        return true;
    }
}

